client/fingerprint/consul.go

```diff
@@ -298,6 +303,60 @@ func (cfs *consulFingerprintState) grpcTLSPort(info agentconsul.Self) (string, b
 	return fmt.Sprintf("%d", int(p)), ok
 }
 
+func (cfs *consulFingerprintState) dnsPort(info agentconsul.Self) (string, bool) {
+	p, ok := info["DebugConfig"]["DNSPort"].(float64)
+	return fmt.Sprintf("%d", int(p)), ok
+}
+
+// dnsAddr fingerprints the Consul DNS address, but only if Nomad can use it
+// usefully to provide an iptables rule to a task
+func (cfs *consulFingerprintState) dnsAddr(logger hclog.Logger) func(info agentconsul.Self) (string, bool) {
+	return func(info agentconsul.Self) (string, bool) {
+
+		// only addresses we can use for an iptables rule from a container to the
+		// host will be fingerprinted
+		isValidForTaskUse := func(addr netip.Addr) (string, bool) {
+			if !addr.IsLoopback() && !addr.IsUnspecified() && addr.IsValid() {
+				return addr.String(), true
+			}
+			return "", false
+		}
+
+		// first try to find an explicitly configured address
+		dnsAddrs, ok := info["DebugConfig"]["DNSAddrs"].([]string)
+		if ok {
+			for _, dnsAddr := range dnsAddrs {
+				dnsAddr = strings.TrimPrefix(dnsAddr, "tcp://")
+				dnsAddr = strings.TrimPrefix(dnsAddr, "udp://")
+
+				parsed, err := netip.ParseAddrPort(dnsAddr)
+				if err != nil {
+					logger.Warn("could not parse Consul addresses.dns config", "value", dnsAddr)
+					return "", false
+				}
+				if val, ok := isValidForTaskUse(parsed.Addr()); ok {
+					return val, true
+				}
+			}
+		}
+
+		// fallback to the bind address
+		bindAddr, ok := info["DebugConfig"]["BindAddr"].(string)
+		if ok {
+			parsed, err := netip.ParseAddr(bindAddr)
+			if err != nil {
+				logger.Warn("could not parse Consul bind_addr config", "value", bindAddr)
+				return "", false
```

Comment:
this is the fallback, but again is `false` what we want? or maybe down the bottom on 356, would we want `true` there, if no value could be found?

I'm just trying to square these with the docstring on `consulExtractor`

> // consulExtractor is used to parse out one attribute from consulInfo. Returns
> // the value of the attribute, and whether the attribute exists.

in this case, since this is the last fallback attempt, `false` seems appropriate since we could not find a value.  but then the final return below has me a bit confused.

Reply:
It's a little subtle and please push back if you disagree, but I'm trying to differentiate between "Consul doesn't give us a DNS address" and "Consul gave us a malformed address, which suggests something's very wrong and we shouldn't start the client"

Reply:
that does seem a good thing to distinguish, and excellent to log as you've done here, but is that distinction meaningful elsewhere that considers "whether the attribute exists"? mainly farther down, does `return "", true` indicating that "the value exists, but it is an empty string" make sense for later use, when actually we have failed to find a value?

pardon my ignorance on the nuances.

Reply:
Yeah, the more I think about it, introducing this subtlety is a bug waiting to happen. We're using a portion of the Consul API response which is marked "unstable" but in practice has been unchanged for a very long time (and in a sidebar with the Consul folks we decided that's fine). If this changes in any way, we should return `"", false` which will bubble up as an error so that we catch it in E2E testing before it ships.

So lemme fix that up and make sure it's always `"", false` if it can't be read.